File: python-to-c++/cc/csvinteresting.cc
#include <fstream>
#include <vector>
#include <string>
#include <regex>
#include <functional>
#include <map>
#include <iostream>
#include <optional>

class CSVReader {
public:
    CSVReader(const std::string& file_path) : file_path(file_path) {
        process_csv();
    }

    void process_csv() {
        std::ifstream csv_file(file_path);
        std::string line;
        if (std::getline(csv_file, line)) {
            column_headings = split(line, ',');
        }
        while (std::getline(csv_file, line)) {
            data.push_back(split(line, ','));
        }
    }

    std::map<std::string, std::string> summary() {
        std::map<std::string, std::string> summary;
        summary["file_path"] = file_path;
        summary["num_rows"] = std::to_string(data.size());
        summary["num_columns"] = std::to_string(column_headings.size());
        summary["column_headings"] = join(column_headings, ",");
        return summary;
    }

    void filter_by_columns(const std::vector<std::string>& columns_to_filter) {
        std::vector<int> column_indices;
        for (const auto& column : columns_to_filter) {
            auto it = std::find(column_headings.begin(), column_headings.end(), column);
            if (it == column_headings.end()) {
                throw std::invalid_argument("Column '" + column + "' not found in CSV file.");
            }
            column_indices.push_back(std::distance(column_headings.begin(), it));
        }
        column_headings = columns_to_filter;
        for (auto& row : data) {
            std::vector<std::string> new_row;
            for (int index : column_indices) {
                new_row.push_back(row[index]);
            }
            row = new_row;
        }
    }

    /// @brief filters by the column. this mutates the column headings and data. so be careful. this might lead to search bugs.
    /// @param search_term 
    /// @param include_columns 
    void contains_filter(const std::string& search_term, const std::vector<std::string>& include_columns) {
        std::vector<std::string> columns_to_filter;
        for (const auto& column_heading : column_headings) {
            if (column_heading.find(search_term) != std::string::npos) {
                columns_to_filter.push_back(column_heading);
            }
        }
        columns_to_filter.insert(columns_to_filter.end(), include_columns.begin(), include_columns.end());
        filter_by_columns(columns_to_filter);
    }

    /// @brief filters by the column. this mutates the column headings and data. so be careful. this might lead to search bugs.
    /// @param prefix 
    /// @param include_columns 
    void prefix_match_filter(const std::string& prefix, const std::vector<std::string>& include_columns) {
        std::vector<std::string> columns_to_filter;
        for (const auto& column_heading : column_headings) {
            if (column_heading.rfind(prefix, 0) == 0) {
                columns_to_filter.push_back(column_heading);
            }
        }
        columns_to_filter.insert(columns_to_filter.end(), include_columns.begin(), include_columns.end());
        filter_by_columns(columns_to_filter);
    }

    /// @brief filters by the column. this mutates the column headings and data. so be careful. this might lead to search bugs.
    /// @param pattern 
    /// @param include_columns 
    void regex_filter(const std::string& pattern, const std::vector<std::string>& include_columns) {
        std::regex re(pattern);
        std::vector<std::string> columns_to_filter;
        for (const auto& column_heading : column_headings) {
            if (std::regex_search(column_heading, re)) {
                columns_to_filter.push_back(column_heading);
            }
        }
        columns_to_filter.insert(columns_to_filter.end(), include_columns.begin(), include_columns.end());
        filter_by_columns(columns_to_filter);
    }

    std::vector<std::string> get_column(const std::string& column_name) {
        std::vector<std::string> column;
        auto it = std::find(column_headings.begin(), column_headings.end(), column_name);
        if (it == column_headings.end()) {
            throw std::invalid_argument("Column '" + column_name + "' not found in CSV file.");
        }
        int column_index = std::distance(column_headings.begin(), it);
        for (const auto& row : data) {
            column.push_back(row[column_index]);
        }
        return column;
    }

    std::vector<std::vector<std::string>> search_column(const std::string& column_name, const std::string& search_term) {
        std::vector<std::vector<std::string>> result;
        auto it = std::find(column_headings.begin(), column_headings.end(), column_name);
        if (it == column_headings.end()) {
            throw std::invalid_argument("Column '" + column_name + "' not found in CSV file.");
        }
        int column_index = std::distance(column_headings.begin(), it);
        for (const auto& row : data) {
            if (row[column_index].find(search_term) != std::string::npos) {
                result.push_back(row);
            }
        }
        return result;
    }

    std::vector<std::vector<std::string>> search_column_regex(const std::string& column_name, const std::string& pattern) {
        std::vector<std::vector<std::string>> result;
        std::regex re(pattern);
        auto it = std::find(column_headings.begin(), column_headings.end(), column_name);
        if (it == column_headings.end()) {
            throw std::invalid_argument("Column '" + column_name + "' not found in CSV file.");
        }
        int column_index = std::distance(column_headings.begin(), it);
        for (const auto& row : data) {
            if (std::regex_search(row[column_index], re)) {
                result.push_back(row);
            }
        }
        return result;
    }

    void replace_column_values_regex(const std::string& column_name, const std::string& pattern, const std::string& new_value) {
        std::regex re(pattern);
        auto it = std::find(column_headings.begin(), column_headings.end(), column_name);
        if (it == column_headings.end()) {
            throw std::invalid_argument("Column '" + column_name + "' not found in CSV file.");
        }
        int column_index = std::distance(column_headings.begin(), it);
        for (auto& row : data) {
            if (std::regex_search(row[column_index], re)) {
                row[column_index] = new_value;
            }
        }
    }

    void add_column(const std::string& column_name, const std::string& default_value = "") {
        column_headings.push_back(column_name);
        for (auto& row : data) {
            row.push_back(default_value);
        }
    }

    void delete_column(const std::string& column_name) {
        auto it = std::find(column_headings.begin(), column_headings.end(), column_name);
        if (it == column_headings.end()) {
            throw std::invalid_argument("Column '" + column_name + "' not found in CSV file.");
        }
        int column_index = std::distance(column_headings.begin(), it);
        column_headings.erase(it);
        for (auto& row : data) {
            row.erase(row.begin() + column_index);
        }
    }

    void rename_column(const std::string& old_name, const std::string& new_name) {
        auto it = std::find(column_headings.begin(), column_headings.end(), old_name);
        if (it == column_headings.end()) {
            throw std::invalid_argument("Column '" + old_name + "' not found in CSV file.");
        }
        *it = new_name;
    }
    std::vector<std::vector<std::string>> data;
    std::vector<std::string> column_headings;

private:
    std::string file_path;

    std::vector<std::string> split(const std::string &str, char delimiter) {
        std::vector<std::string> tokens;
        std::string token;
        for (char c : str) {
        if (c == delimiter) {
            tokens.push_back(token);
            token.clear();
        } else {
            token += c;
        }
        }
        // Add the last token if it's not empty
        if (!token.empty()) {
           tokens.push_back(token);
        }
        return tokens;
    }

    std::string join(const std::vector<std::string>& tokens, std::string delimiter) {
        std::string str;
        for (size_t i = 0; i < tokens.size(); ++i) {
            str += tokens[i];
            if (i != tokens.size() - 1) {
                str += delimiter;
            }
        }
        return str;
    }
};


int main() {
    // read the data at world-cup-2022/wc_forecasts.csv
    CSVReader csv_reader("world-cup-2022/wc_forecasts.csv");

    // print the column headings
    for (const auto& column_heading : csv_reader.column_headings) {
        std::cout << column_heading << ' ';
    }
    std::cout << std::endl;

    return 0;
}

File: python-to-c++/py/csvinteresting.py
import csv
import re
import types

class CSVReader:
    def __init__(self, file_path):
        self.file_path = file_path
        self.data = []
        self.column_headings = []
        self.process_csv()

        self.predefined_column_filters = {
            'contains': self.contains_filter,
            'regex': self.regex_filter,
            'prefix-match': self.prefix_match_filter,
        }

    def process_csv(self):
        with open(self.file_path, 'r') as csv_file:
            csv_reader = csv.reader(csv_file)
            self.column_headings = next(csv_reader)
            for row in csv_reader:
                self.data.append(row)

    def summary(self):
        num_rows = len(self.data)
        num_columns = len(self.column_headings)
        return {
            'file_path': self.file_path,
            'num_rows': num_rows,
            'num_columns': num_columns,
            'column_headings': self.column_headings
        }

    """
    Useful Utils
    """
    def filter_by_columns(self, columns_to_filter):
        """
        Filters the data to only include the specified columns.

        Args:
            columns_to_filter: List of column names to filter by.

        Raises:
            ValueError: If columns_to_filter is not a list.
            ValueError: If any column name in columns_to_filter is not found in the CSV file.
        """
        # Sanity check: columns_to_filter must be a list
        if not isinstance(columns_to_filter, list):
            raise ValueError("columns_to_filter must be a list of column names.")

        # Sanity check: all column names in columns_to_filter must be found in the CSV file
        for column in columns_to_filter:
            if column not in self.column_headings:
                raise ValueError(f"Column '{column}' not found in CSV file.")

        # Filter the data to only include the specified columns
        column_indices = [self.column_headings.index(column) for column in columns_to_filter]
        self.column_headings = columns_to_filter
        self.data = [[row[index] for index in column_indices] for row in self.data]

    """
    Predefined filters
    """
    def contains_filter(self, search_term, include_columns):
        """filter out the column that contains the search term"""
        columns_to_filter = []
        for column_heading in self.column_headings:
            if search_term in column_heading:
                columns_to_filter.append(column_heading)
        return self.filter_by_columns(columns_to_filter + include_columns)

    def prefix_match_filter(self, prefix, include_columns):
        """Filters the data to only include columns that start with the specified prefix."""
        columns_to_filter = []
        for column_heading in self.column_headings:
            if column_heading.startswith(prefix):
                columns_to_filter.append(column_heading)
        self.filter_by_columns(columns_to_filter + include_columns)

    def regex_filter(self, pattern, include_columns):
        """Filters the data to only include columns that match the regex pattern."""
        columns_to_filter = []
        for column_heading in self.column_headings:
            if re.search(pattern, column_heading):
                columns_to_filter.append(column_heading)
        self.filter_by_columns(columns_to_filter + include_columns)

    """
    Column operations
    """
    def get_column(self, column_name):
        """Returns a list of values from the specified column."""
        column_index = self.column_headings.index(column_name)
        return [row[column_index] for row in self.data]

    def search_column(self, column_name, search_term):
        """Returns rows where the specified column contains the search term."""
        column_index = self.column_headings.index(column_name)
        return [row for row in self.data if search_term in row[column_index]]

    def search_column_regex(self, column_name, pattern):
        """Returns rows where the specified column matches the regex pattern."""
        column_index = self.column_headings.index(column_name)
        return [row for row in self.data if re.search(pattern, row[column_index])]

    def replace_column_values_regex(self, column_name, pattern, new_value):
        """Replaces all values in the specified column that match the regex pattern with new_value."""
        column_index = self.column_headings.index(column_name)
        for row in self.data:
            if re.search(pattern, row[column_index]):
                row[column_index] = new_value

    def add_column(self, column_name, default_value=None):
        """Adds a new column with the specified name and fills it with the default value."""
        self.column_headings.append(column_name)
        for row in self.data:
            row.append(default_value)

    def delete_column(self, column_name):
        """Deletes the specified column."""
        column_index = self.column_headings.index(column_name)
        self.column_headings.pop(column_index)
        for row in self.data:
            row.pop(column_index)

    def rename_column(self, old_name, new_name):
        """Renames the specified column."""
        column_index = self.column_headings.index(old_name)
        self.column_headings[column_index] = new_name

    """
    Custom Operations
    """
    def group_by_column_and_filter(self, column_name, filter_tuple=None, aggregation_function=None):
        """
        Groups the data by the specified column, filters the columns, and aggregates each group independently.

        Creates the following files:
            1. One CSV file per group, named after the group value.
            2. One CSV file containing the aggregated data per group.

        Args:
            column_name: The column to group by.
            filter_tuple: A tuple with the (name of the filter function, its argument).
            aggregation_function: The function to use to aggregate the values in each group. Optional.
        """

        # Sanity check: filter_tuple must be a string
        if filter_tuple is not None:
            if not isinstance(filter_tuple, tuple) or len(filter_tuple) != 2:
                raise ValueError("filter_tuple must be a tuple of length 2 to filter by.")

        # First filter the columns
        if filter_tuple is not None:
            filter_func = self.predefined_column_filters[filter_tuple[0]]
            filter_func(filter_tuple[1], [column_name])

        # Sanity check: column_name must be found in the CSV file
        if column_name not in self.column_headings:
            raise ValueError(f"Column '{column_name}' not found in CSV file.")


        # Group the data by the specified column
        column_index = self.column_headings.index(column_name)
        groups = {}
        for row in self.data:
            group_key = row[column_index]

            if group_key not in groups:
                groups[group_key] = []
            groups[group_key].append(row)

        # Sanity check: aggregation_function must be a function
        if aggregation_function is not None:
            if not isinstance(aggregation_function, types.FunctionType):
                raise ValueError("aggregation_function must be a function.")

        # Aggregate the values in each group
        aggregations = {}
        if aggregation_function is not None:
            for group_key, group_data in groups.items():
                aggregations[group_key] = aggregation_function(group_data)

        # Make a new CSV file for each group
        for group_key, group_data in groups.items():
            file_path = f"{self.file_path.split('.')[0]}_{group_key}.csv"
            with open(file_path, 'w') as csv_file:
                csv_writer = csv.writer(csv_file)
                csv_writer.writerow(self.column_headings)
                csv_writer.writerows(group_data)

        # Write the aggregated data to a CSV file
        file_path = f"{self.file_path.split('.')[0]}_aggregations.csv"
        with open(file_path, 'w') as csv_file:
            csv_writer = csv.writer(csv_file)
            csv_writer.writerow([column_name, 'Aggregation'])
            for group_key, aggregation in aggregations.items():
                csv_writer.writerow([group_key, aggregation])


if __name__ == '__main__':
    # read the data at world-cup-2022/wc_forecasts.csv

    csv_reader = CSVReader('world-cup-2022/wc_forecasts.csv')

    # print the column headings
    print(csv_reader.column_headings)

    # filter so that it contains sim
    csv_reader.group_by_column_and_filter('group', ("regex", '(sim)|(team)'), lambda x: len(x))